### PR TITLE
Rule S6613: implement CodeFix

### LINKED ISSUE #7239: Rule S6613: implement CodeFix
Implement a CodeFix for S6613.

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #7239
RSPEC PR: https://github.com/SonarSource/rspec/pull/1953

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(LinkedListPropertiesInsteadOfMethods.DiagnosticId);
+
+        protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+
+            if (root.FindNode(diagnosticSpan, getInnermostNodeForTie: true) is IdentifierNameSyntax identifierSyntax
+                && identifierSyntax is { Parent: ExpressionSyntax { Parent: InvocationExpressionSyntax invocationExpression } expression })
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs:39]: Parent `ExpressionSyntax` will cover both `SimpleMemberAccessExpression` for plain invocations and `MemberBindingExpression` for conditional expressions.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public sealed class LinkedListPropertiesInsteadOfMethodsCodeFix : SonarCodeFix
+    {
+        internal const string Title = "Use the 'LinkedList' property";
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs:30]: pedantic: you can insert an empty line between internal fields and properties.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+            var identifierSyntax = (IdentifierNameSyntax)root.FindNode(diagnosticSpan, getInnermostNodeForTie: true);
+
+            if (identifierSyntax is { Parent: ExpressionSyntax { Parent: InvocationExpressionSyntax invocationExpression } expression })
+            {
+                var newMember = SyntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expression, SyntaxFactory.IdentifierName("Value"));
+                context.RegisterCodeFix(
+                    CodeAction.Create(
+                        Title,
+                        c =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs:45]: pedantic: `c` is not used and we can use the discard pattern

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public sealed class LinkedListPropertiesInsteadOfMethodsCodeFix : SonarCodeFix
+    {
+        internal const string Title = "Use the 'LinkedList' property";
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs:29]: the `Title` can be private

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public sealed class LinkedListPropertiesInsteadOfMethodsCodeFix : SonarCodeFix
+    {
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LinkedListPropertiesInsteadOfMethodsCodeFix.cs:28]: The way the title is now, makes me think that the code-fix will replace the previous call with `LinkedList`. I might be wrong but wouldn't be better to have instead: "Replace extension method call with property"? Or something similar. We can ask for help from a native speaker.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LinkedListPropertiesInsteadOfMethods.cs [hunk 6]
```diff
@@ -7,10 +7,10 @@ class MyClass
 {
     void Basic(LinkedList<int> data)
     {
-        data.First(); // Noncompliant {{'First' property of 'LinkedList' should be used instead of the 'First()' extension method.}}
-//           ^^^^^
-        data.Last(); // Noncompliant {{'Last' property of 'LinkedList' should be used instead of the 'Last()' extension method.}}
-//           ^^^^
+        _ = data.First(); // Noncompliant {{'First' property of 'LinkedList' should be used instead of the 'First()' extension method.}}
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LinkedListPropertiesInsteadOfMethods.cs:10]: A lot of the calls (not all) have been changed to be assignments. Can I ask what's the reason for this change?

**Review by costin-zaharia-sonarsource — COMMENTED** [review]
It looks good to me. I've added a few pedantic comments that are optional and a question to better understand why some of the test cases were changed.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LinkedListPropertiesInsteadOfMethods.cs (hunk 6 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LinkedListPropertiesInsteadOfMethods.cs:10]: So, in some cases the extension method was called without an assignment to improve the readability:
`data.First();`
The problem arises when we fix that issue with the LinkedList property:
`data.First.Value;`
This will raise a `CS0201: Only assignment, call, increment, decrement, await, and new object expressions can be used as a statement`. That's why I added the assignment with the discard variable to the invalid statements.

**Review by sebastien-marichal — APPROVED** [review]
LGTM!

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! The failing test seems to be related to other changes and not this PR. Can you please check before merging?
